### ISSUE TITLE
CBSE Maths grade 10 AI/ML activities

### ISSUE THREAD

**Issue by ndubbaka** [Pragament/coding_simulator_projects#8]
Tools
- https://machinelearningforkids.co.uk/scratch/
- https://teachablemachine.withgoogle.com/
- https://playground.raise.mit.edu/create/
- https://quickdraw.withgoogle.com/
- https://experiments.withgoogle.com/thing-translator
- https://soundraw.io/generate_music
- https://research.aimultiple.com/generative-ai-tools/
- https://www.freshbots.org/
- https://appinventor.mit.edu/

**For Class 10 students**, we can introduce **production-level AI/ML concepts** aligned with their mathematical sophistication and impending career choices. Here are **advanced AI/ML activities tailored to Class 10 topics**:

---

## **🧠 AI Concepts for Class 10 Level:**
- **Deep learning fundamentals**
- **Convolutional Neural Networks (CNNs)**
- **Training/Validation/Test workflow**
- **Hyperparameter tuning**
- **Ensemble methods**
- **Model deployment concepts**
- **Ethical AI & bias detection**

---

### **Activity: "Find Roots without Formula"**
1. **Problem:** Solve ax² + bx + c = 0 using ML approach
2. **Reformulate as Optimization:**
   - Instead of formula, minimize f(x) = (ax² + bx + c)²
   - Roots occur where f(x) = 0 (minimum)
3. **Gradient Descent Implementation:**
   ```python (conceptual)
   # Initialize two x values (for two roots)
   x1, x2 = random(), random()
   # Update rule: x = x - learning_rate * gradient
   gradient = 2*(ax² + bx + c)*(2ax + b)
   ```
4. **Visualization:** Show GD finding each root
5. **Compare:** With discriminant method (b² - 4ac analysis)
6. **Edge Cases:** Complex roots handling

### **Extension:** Use TensorFlow/PyTorch autograd

---

### **Activity: "Next Number Predictor"**
1. **Dataset Generation:** Create AP sequences with noise
2. **Feature Engineering:**
   - First differences (constant for AP)
   - Second differences (zero for AP)
   - Ratio of terms (for GP detection)
3. **Model Architecture:**
   - **Simple:** Linear regression on position index
   - **Advanced:** RNN/LSTM for sequence prediction
4. **Training:** Predict next 3 terms given first 5
5. **Evaluation Metrics:**
   - Mean Absolute Error (MAE)
   - Direction accuracy (increasing/decreasing)

### **Real Application:** Stock trend prediction (simplified)

---

### **Activity: "Circle Property Detector"**
1. **Dataset Creation:** Generate circle images with:
   - Tangent lines
   - Chords
   - Secants
   - Cyclic quadrilaterals
2. **CNN Architecture:**
   - Conv layers detect edges, curves
   - Pooling layers identify circle parts
   - Dense layers classify properties
3. **Tasks:**
   - Detect if line is tangent
   - Find angle in alternate segment
   - Identify equal chords
4. **Visualization:** Show what CNN "sees" (activation maps)

### **Tool:** Teachable Machine (advanced) or FastAI

---

### **Activity: "Build sin(x) using Neurons"**
1. **Problem:** Create neural network that approximates sin(x)
2. **Network Design:**
   - Input: x (angle in radians)
   - Hidden layers: 2-3 layers with activation functions
   - Output: sin(x) prediction
3. **Training Data:** x from 0 to 2π, y = sin(x)
4. **Comparison Methods:**
   - Polynomial approximation (Taylor series)
   - Lookup table interpolation
   - Neural network approach
5. **Application:** Height calculation from angle and distance

### **Mathematical Insight:** Fourier analysis connection

---

### **Activity: "Optimal Shape Design"**
1. **Optimization Problem:** Given fixed volume, minimize surface area (or vice versa)
2. **Genetic Algorithm Components:**
   - **Gene:** Dimensions (radius, height)
   - **Fitness:** Surface area (to minimize)
   - **Crossover:** Mix parent dimensions
   - **Mutation:** Random dimension changes
3. **Evolution Simulation:**
   - Generation 1: Random shapes
   - Generation 10: Better shapes emerging
   - Generation 50: Near-optimal sphere/cylinder
4. **Mathematical Verification:** Compare with calculus solutions

### **Real World:** Engineering design optimization

---

### **Activity: "Geometric Theorem Prover"**
1. **Problem:** Given coordinates, prove geometric properties
2. **Representation:**
   - Points as tokens (A, B, C)
   - Coordinates as embeddings
   - Relationships as attention weights
3. **Attention Mechanism:**
   - Which points are related? (Self-attention)
   - How are they related? (Cross-attention)
4. **Tasks:**
   - Prove collinearity
   - Find intersection points
   - Calculate area
5. **Visualization:** Attention maps showing "focus"

### **Connection:** Google's AlphaGeometry

---

### **Activity: "Spam Filter using Bayes Theorem"**
1. **Mathematical Foundation:** 
   - P(Spam|Word) = P(Word|Spam)P(Spam) / P(Word)
2. **Dataset:** Simple email words (FREE, WIN, MEETING, HOMEWORK)
3. **Training:** Count word frequencies in spam/ham
4. **Prediction:** New email → calculate spam probability
5. **Evaluation:**
   - Precision/Recall tradeoff
   - Confusion matrix analysis
6. **Assumption Check:** Why "Naive"? (Word independence)

### **Extension:** Sentiment analysis on math problem statements

---

### **Activity: "Predict Class Performance"**
1. **Dataset:** Student features + final scores
2. **Feature Engineering:**
   - Previous test averages
   - Attendance percentage
   - Homework completion
   - Study hours
3. **Random Forest Construction:**
   - Build multiple decision trees on random data subsets
   - Each tree makes prediction
   - Final prediction = average of all trees
4. **Advantages:** 
   - Reduces overfitting
   - Handles missing data better
   - Feature importance scores
5. **Interpretation:** Which factors most affect scores?

### **Tool:** Google Colab with scikit-learn RandomForest

---

### **Activity: "Linear Regression as Matrix Solution"**
1. **Matrix Formulation:**
   ```
   y = Xβ + ε
   Normal equation: β = (XᵀX)⁻¹Xᵀy
   ```
2. **Geometric Interpretation:** 
   - X columns span feature space
   - y projection onto this space
3. **Numerical Issues:** 
   - When XᵀX is singular (multicolinearity)
   - Ridge regression solution: β = (XᵀX + λI)⁻¹Xᵀy
4. **Comparison:** Gradient descent vs matrix solution
5. **Application:** Predict marks in two subjects

### **Advanced:** Singular Value Decomposition (SVD) demo

---

### **Activity: "Build a Math Tutor AI System"**
1. **Problem Definition:** Identify student's weak areas
2. **Data Collection:** Past performance, question attempts
3. **Feature Engineering:** 
   - Topic difficulty scores
   - Time spent patterns
   - Error type classification
4. **Model Selection:** 
   - Classification: Topic difficulty
   - Regression: Score prediction
   - Recommendation: Next practice problems
5. **Deployment:** Simple web interface
6. **Monitoring:** Track model performance over time

### **Tools:** Streamlit for frontend, scikit-learn for ML

---

## **🛠️ Production-Ready Tools for Class 10:**

1. **Google Colab Pro** (GPU access for training)
2. **TensorFlow/Keras** (deep learning frameworks)
3. **Streamlit/Flask** (web deployment)
4. **Git/GitHub** (version control)
5. **Docker basics** (containerization)
6. **MLflow** (experiment tracking)

---

### **Project 1: "Intelligent Exam Paper Generator"**
1. **Input:** Syllabus, difficulty level, topic weights
2. **ML Tasks:**
   - Question difficulty prediction
   - Topic coverage optimization
   - Similar question detection
3. **Output:** Balanced question paper
4. **Evaluation:** Teacher feedback loop

### **Project 2: "Automated Geometry Diagram Analyzer"**
1. **Input:** Hand-drawn geometry diagram (photo)
2. **Processing:**
   - CNN for shape recognition
   - OCR for labels
   - Geometric constraint solver
3. **Output:** Properties, proofs, area calculations
4. **Tech Stack:** OpenCV + TensorFlow + SymPy

### **Project 3: "Personalized Learning Path Optimizer"**
1. **Data:** Student learning patterns
2. **Reinforcement Learning Approach:**
   - State: Current knowledge
   - Action: Next topic/exercise
   - Reward: Learning gain
3. **Goal:** Optimal sequence for fastest mastery
4. **Validation:** A/B testing with student groups

---

### **1. "Attention in Geometric Proofs"**
- How transformers can prove geometric theorems
- Attention visualization in AlphaGeometry
- Limitations of current AI in pure math

### **2. "Differentiable Rendering for Math"**
- Convert diagrams to differentiable representations
- Gradient-based optimization of geometric parameters
- Applications in CAD/CAM

### **3. "Neuro-Symbolic AI for Math"**
- Combine neural networks with symbolic logic
- Neural part: Pattern recognition
- Symbolic part: Rule application
- Best of both worlds

---

### **Comprehensive Metrics:**
1. **Regression:** MAE, MSE, R², Adjusted R²
2. **Classification:** Precision, Recall, F1, ROC-AUC
3. **Clustering:** Silhouette score, Davies-Bouldin index
4. **Time Series:** MAPE, SMAPE, MASE

### **Validation Techniques:**
- k-Fold Cross-validation
- Time-series cross-validation
- Stratified sampling for imbalanced data

---

### **Critical Discussions:**
1. **Bias in Educational AI:**
   - If trained on urban elite data, fails for rural students
   - Gender bias in math problem selection
   - Socioeconomic bias in performance prediction

2. **Transparency & Explainability:**
   - Why did AI recommend this topic?
   - SHAP/LIME for model interpretation
   - Right to explanation

3. **Data Privacy:**
   - Student data protection
   - Anonymous data collection
   - GDPR compliance basics

### **Activity: "Audit a Math AI System"**
- Given a hypothetical AI tutor, identify potential biases
- Suggest fairness improvements
- Design ethical guidelines

---

### **Phase 1: Foundation (2 months)**
- Python programming basics
- NumPy, Pandas for data manipulation
- Basic statistics and visualization

### **Phase 2: Core ML (3 months)**
- Supervised learning algorithms
- Model evaluation techniques
- Feature engineering

### **Phase 3: Advanced Topics (3 months)**
- Neural networks introduction
- Project work
- Ethics and deployment

### **Phase 4: Capstone (2 months)**
- End-to-end project
- Documentation and presentation
- Portfolio development

---

### **Hackathon Ideas:**
1. **"Math Accessibility Challenge"**: AI tools for visually impaired
2. **"Rural Education Innovation"**: Low-resource AI solutions
3. **"Cross-cultural Math AI"**: Multiple language/curriculum support

### **Portfolio Projects:**
- GitHub repository with documented code
- Live demo (Streamlit/Heroku)
- Technical report (Google Docs/Overleaf)
- Video presentation

---

## **🔗 Industry Connections:**

1. **Research Papers:** Simplified versions of:
   - "Attention Is All You Need" (Transformers)
   - "ImageNet Classification with Deep CNN" (AlexNet)
   - "Playing Atari with Deep RL" (DeepMind)

2. **Industry Tools:**
   - Google's Vertex AI
   - Microsoft Azure ML
   - Amazon SageMaker (conceptual)

3. **Open Source Contributions:**
   - Documentation improvements
   - Simple bug fixes
   - Dataset contributions

---

### **Technical (60%):**
- Code quality and documentation
- Model performance metrics
- Innovative approach

### **Communication (20%):**
- Clear explanations
- Effective visualizations
- Concise documentation

### **Ethics & Impact (20%):**
- Bias analysis
- Privacy considerations
- Real-world applicability

### **Final Deliverables:**
1. Working prototype
2. Technical report
3. 5-minute demo video
4. Reflection on ethical considerations

---

## **🚨 Real-World Precautions:**

1. **Data Privacy:** Never use real student names/sensitive data
2. **Realistic Expectations:** AI as assistant, not replacement
3. **Academic Integrity:** Clear boundaries for AI assistance
4. **Accessibility:** Ensure solutions work for diverse learners

---

These activities transform **Class 10 students from consumers to creators of AI**, equipping them with **industry-relevant skills** while deepening their **mathematical understanding** through practical application!
